usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +491,394 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+            return CKR_MECHANISM_INVALID;
+        }
+        rc = check_ab_kek_type(wrap_key_obj->template, CKR_WRAPPING_KEY_TYPE_INCONSISTENT);
+        if (rc != CKR_OK)
+            return rc;
+        rc = h_opaque_2_blob(tokdata,
+                             ((CK_IBM_ATTRIBUTEBOUND_WRAP_PARAMS *)mech->pParameter)->hSignVerifyKey,
+                             sblob, sblob_len, sobj, READ_LOCK);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("Failed to get sign key for AB wrapping\n");
+            return rc;
+        }
+        rc = template_attribute_get_bool((*sobj)->template, CKA_IBM_ATTRBOUND, &absignkey);
+        if (rc != CKR_OK || !absignkey) {
+            TRACE_ERROR("AB-Wrap: sign key not AB\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto out;
+        }
+        rc = template_attribute_get_bool((*sobj)->template, CKA_SIGN, &signsignkey);
+        if (rc != CKR_OK || !signsignkey) {
+            TRACE_ERROR("AB-Wrap: sign key not able to sign\n");
+            rc = CKR_KEY_FUNCTION_NOT_PERMITTED;
+            goto out;
+        }
+    } else if (abkey) {
+        TRACE_ERROR("Target key is AB, but wrapping key is not\n");
+        return CKR_KEY_FUNCTION_NOT_PERMITTED;
+    }
+ out:
+    if (rc != CKR_OK && *sobj) {
+        object_put(tokdata, *sobj, TRUE);
+        *sobj = 0;
+        *sblob = 0;
+        *sblob_len = ~0;
+    }
+    return rc;
+}
+
+static CK_RV check_ab_unwrap(STDLL_TokData_t * tokdata,
+                             CK_BYTE **vblob, size_t *vblob_len,
+                             OBJECT **vobj, OBJECT *unwrap_key_obj,
+                             CK_ATTRIBUTE_PTR attrs, CK_ULONG attrs_len,
+                             CK_MECHANISM_PTR mech, CK_BBOOL *isab)
+{
+    CK_RV rc;
+    CK_BBOOL abkey, abunwrapkey, abverifykey, verifyverifykey;
+
+    *vobj = 0;
+    *vblob = 0;
+    *vblob_len = ~0;
+    *isab = FALSE;
+    rc = template_attribute_get_bool(unwrap_key_obj->template, CKA_IBM_ATTRBOUND,
+                                     &abunwrapkey);
+    if (rc == CKR_TEMPLATE_INCOMPLETE)
+        abunwrapkey = FALSE;
+    else if (rc != CKR_OK)
+        return rc;
+    rc = get_bool_attribute_by_type(attrs, attrs_len, CKA_IBM_ATTRBOUND, &abkey);
+    if (rc == CKR_TEMPLATE_INCOMPLETE)
+        abkey = FALSE;
+    else if (rc != CKR_OK)
+        return rc;
+    if (mech->mechanism == CKM_IBM_ATTRIBUTEBOUND_WRAP) {
+        *isab = TRUE;
+        if (mech->ulParameterLen != sizeof(CK_IBM_ATTRIBUTEBOUND_WRAP_PARAMS)) {
+            TRACE_ERROR("Unwrapping AB key without CKM_IBM_ATTRIBUTEBOUND_WRAP\n");
+            return CKR_MECHANISM_INVALID;
+        }
+        if (!abkey) {
+            TRACE_ERROR("CKM_IBM_ATTRIBUTEBOUND_WRAP with non-AB target\n");
+            return CKR_MECHANISM_INVALID;
+        }
+        if (!abunwrapkey) {
+            TRACE_ERROR("CKM_IBM_ATTRIBUTEBOUND_WRAP with non-AB unwrap key\n");
+            return CKR_KEY_FUNCTION_NOT_PERMITTED;
```

Comment:
Same question as above, although CKR_KEY_FUNCTION_NOT_PERMITTED looks like a good choice. Maybe also use CKR_KEY_FUNCTION_NOT_PERMITTED  for above?

Reply:
I see this as a different situation.  Here, we want to unwrap into an AB key with a non-AB unwrapping key and CKM_IBM_ATTRIBUTEBOUND_WRAP.  So the unwrapping key is here the problem.  I was thinking about CKR_UNWRAPPING_KEY_TYPE_INCONSISTENT but decided against it because it is not the key type, but the key function "AB unwrap" that is not permitted/supported by this key.

Reply:
Yes, keep CKR_KEY_FUNCTION_NOT_PERMITTED it fits well.